### ISSUE TITLE
[Feature Request] Handle edge case of recursive exceptions in failure converter

### ISSUE THREAD

**Issue by yunmanger1** [temporalio/sdk-python#697]
### Is your feature request related to a problem? Please describe.

It is kind of an easy mistake to do in python:

```
try:
    raise ValueError("test")
except Exception as e:
    raise e from e
```

The re-raised exception will have `e.__cause__` referencing itself resulting in failure to serialize error and python sdk won't report anything resulting in retries and startToClose timeouts.

### Describe the solution you'd like

Handle a case  when `e.__cause__ == e` in DefaultFailureConverter. Created a dummy pull request #696

### Additional context

https://temporalio.slack.com/archives/CTT84RS0P/p1719632201314119

**Comment by cretz** [comment]
> python sdk won't report anything

I think this is the primary bug. There are situations where recursion and stack overflow can occur in serialization, and the linked PR naively only helps the case of the cause being the same as itself (but if it was two causes deep that recursed, the issue would still appear).

If we are indeed not reporting anything, we need to fix that. #685 is related when failing to build a failure.

**Comment by yunmanger1** [comment]
You are right that it doesn't solve all cycles. Not sure if your case happens in practice. What needs to happen to have an exception cycled after 2 or 3 steps. 

Maybe the solution could be just have a max depth of 10 let's say. During serialization you truncate too  deep stacks anyway.

**Comment by cretz** [comment]
I think we just need to log the failure of serializing this (assume there's a stack depth/overflow exception but is not visible due to #685)

**Comment by Ye123459** [comment]
maybe  you can take the following solutions:

Avoid using the from statement when catching and rethrowing exceptions:

Reraising the exception directly without specifying from avoids the problem of __cause__ pointing to itself.
Check and reset the __cause__ attribute:

After catching the exception, check the __cause__ attribute and reset it to None if it points to itself.
Custom exception handling logic:

Add logic to DefaultFailureConverter to ensure that __cause__ does not point to itself when handling exceptions.
Here is a specific code example:
python
class DefaultFailureConverter:
    def convert(self, exception):
        try:
            # Let's say there's some processing logic here
            raise exception
        except Exception as e:
            #Check if __cause__ refers to itself
            if e.__cause__ is e:
                e.__cause__ = None
            # Reraise the exception, avoiding the from statement
            raise e

**Comment by Ye123459** [comment]
When handling exceptions in DefaultFailureConverter, to avoid e.__cause__ == e problems caused by reraising exceptions, you can take the following solutions:

Avoid using the from statement when catching and rethrowing exceptions:

Reraising the exception directly without specifying from avoids the problem of __cause__ pointing to itself.
Check and reset the __cause__ attribute:

After catching the exception, check the __cause__ attribute and reset it to None if it points to itself.
Custom exception handling logic:

Add logic to DefaultFailureConverter to ensure that __cause__ does not point to itself when handling exceptions.
Here is a specific code example:

class DefaultFailureConverter:
    def convert(self, exception):
        try:
            raise exception
        except Exception as e:
            if e.__cause__ is e:
                e.__cause__ = None
            raise e
